tap_zendesk_chat/streams.py

```diff
@@ -114,93 +99,110 @@ def _pull(self, ctx, chat_type, ts_field, *, full_sync):
+            next_sync = strptime_to_utc(last_sync) + timedelta(days=int(sync_days))
             if next_sync <= ctx.now:
-                LOGGER.info("Running full sync of chats: "
-                            "last sync was %s, configured to run every %s days",
-                            last_sync, sync_days)
+                LOGGER.info(
+                    "Running full sync of chats: last sync was %s, configured to run every %s days",
+                    last_sync,
+                    sync_days,
+                )
                 return True
         return False
 
-    def sync(self, ctx):
+    def sync(self, ctx, schema: Dict, stream_metadata: Dict, transformer: Transformer):
         full_sync = self._should_run_full_sync(ctx)
-        self._pull(ctx, "chat", "end_timestamp", full_sync=full_sync)
-        self._pull(ctx, "offline_msg", "timestamp", full_sync=full_sync)
+        self._pull(
+            ctx,
+            "chat",
+            "end_timestamp",
+            full_sync=full_sync,
+            schema=schema,
+            stream_metadata=stream_metadata,
+            transformer=transformer,
+        )
+        self._pull(
+            ctx,
+            "offline_msg",
+            "timestamp",
+            full_sync=full_sync,
+            schema=schema,
+            stream_metadata=stream_metadata,
+            transformer=transformer,
+        )
         if full_sync:
             ctx.state["chats_last_full_sync"] = ctx.now.isoformat()
             ctx.write_state()
 
 
 class Bans(Stream):
-    def sync(self, ctx):
-        with Transformer() as transformer:
-            schema = ctx.catalog.get_stream(self.tap_stream_id).schema.to_dict()
-            m_data = metadata.to_map(ctx.catalog.get_stream(self.tap_stream_id).metadata)
-            response = ctx.client.request(self.tap_stream_id)
+    def sync(self, ctx, schema: Dict, stream_metadata: Dict, transformer: Transformer):
+        since_id_offset = [self.tap_stream_id, "offset", "id"]
+        since_id = ctx.bookmark(since_id_offset) or 0
+
+        while True:
+            params = {
+                "since_id": since_id,
+                "limit": ctx.config.get("bans_page_limit", 100),
+            }
+            response = ctx.client.request(self.tap_stream_id, params)
+            page = response.get("visitor", []) + response.get("ip_address", [])
+            if not page:
+                break
             page = response["visitor"] + response["ip_address"]
-            page = [transformer.transform(rec, schema, metadata=m_data) for rec in page]
-            self.write_page(page)
+            self.write_page([transformer.transform(rec, schema, metadata=stream_metadata) for rec in page])
+            since_id = page[-1]["id"] + 1
+            ctx.set_bookmark(since_id_offset, since_id)
+            ctx.write_state()
+        ctx.set_bookmark(since_id_offset, None)
+        ctx.write_state()
```

Comment:
is it necessary to write state here since we have same step in sync.py [here](https://github.com/singer-io/tap-zendesk-chat/blob/c435e42065a59f99fbeb0b66ec99e88529c8613a/tap_zendesk_chat/sync.py#L29)

Reply:
removed the duplicate state logs